File: app/src/main/java/ru/taximer/taxiandroid/ui/MainActivity.java
package ru.taximer.taxiandroid.ui;


import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.BottomSheetBehavior;
import android.support.design.widget.BottomSheetDialog;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.View;
import android.view.WindowManager;
import android.widget.FrameLayout;
import android.widget.TextView;
import android.widget.Toast;
import android.widget.RelativeLayout.LayoutParams;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapView;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;

import java.io.IOException;
import java.util.List;
import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;
import ru.taximer.taxiandroid.R;
import ru.taximer.taxiandroid.network.Config;

public class MainActivity extends AppCompatActivity implements OnMapReadyCallback {

    @BindView(R.id.mapView)
    MapView mapView;
    @BindView(R.id.recyclerview)
    RecyclerView autocompleteAddresses;
    @BindView(R.id.drag_me)
    TextView tmp_bar;
//    @BindView(R.id.app_bar)
//    AppBarLayout appbar;
//    @BindView(R.id.toolbar)
//    Toolbar toolbar;

    GoogleMap map;
    BottomSheetBehavior mBottomSheetBehavior;
    LocationManager locationManager;
    Geocoder geocoder;

    private LocationListener mLocationListener = new LocationListener() {

        @Override
        public void onStatusChanged(String provider, int status, Bundle extras) {

        }

        @Override
        public void onProviderEnabled(String provider) {

        }

        @Override
        public void onProviderDisabled(String provider) {

        }

        @Override
        public void onLocationChanged(Location location) {
            if (location != null) {
                Toast.makeText(getApplicationContext(), String.format("%f, %f", location.getLatitude(), location.getLongitude()), Toast.LENGTH_SHORT).show();
                try {
                    drawMarker(location);
                } catch (IOException e) {
                    e.printStackTrace();
                }
                locationManager.removeUpdates(mLocationListener);
            } else {
                Toast.makeText(getApplicationContext(),"Location is null", Toast.LENGTH_SHORT).show();
            }
        }

    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        locationManager = (LocationManager) getSystemService(LOCATION_SERVICE);

        mapView.onCreate(savedInstanceState);
        mapView.getMapAsync(this);

        FrameLayout sheetView = (FrameLayout) autocompleteAddresses.getParent().getParent();
        mBottomSheetBehavior = BottomSheetBehavior.from(sheetView);
        if (mBottomSheetBehavior != null) {
            mBottomSheetBehavior.setBottomSheetCallback(new BottomSheetBehavior.BottomSheetCallback() {
                @Override
                public void onStateChanged(@NonNull View bottomSheet, int newState) {
                }

                @Override
                public void onSlide(@NonNull View bottomSheet, float slideOffset) {
                }
            });
        }

        mBottomSheetBehavior.setPeekHeight(130);
        tmp_bar.requestLayout();

        try {
            getCurrentLocation();
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    @Override
    protected void onResume() {
        super.onResume();
        mapView.onResume();
    }
    @Override
    protected void onPause() {
        super.onPause();
        mapView.onPause();
    }

    @Override
    public void onBackPressed() {
        if (mBottomSheetBehavior.getState() != BottomSheetBehavior.STATE_HIDDEN) {
            mBottomSheetBehavior.setState(BottomSheetBehavior.STATE_COLLAPSED);
        } else {
            super.onBackPressed();
        }
    }


    //TODO перенести это все (с картой) в отдельный класс/презентер
    @Override
    public void onMapReady(GoogleMap googleMap) throws SecurityException { //TODO добавить рантайм проверку
        map = googleMap;
        map.setIndoorEnabled(true);
        map.setMyLocationEnabled(true);
//        map.addMarker(new MarkerOptions().position(new LatLng(40.7143528, -74.0059731)));
//        map.moveCamera(CameraUpdateFactory.newLatLng(new LatLng(40.7143528, -74.0059731)));
        map.moveCamera(CameraUpdateFactory.zoomTo(11));
    }

    private void drawMarker(Location location) throws IOException{
        if (map != null) {
            List<Address> addresses;
            geocoder = new Geocoder(this, Locale.getDefault());
            addresses = geocoder.getFromLocation(location.getLatitude(), location.getLongitude(), 1);
            String address = addresses.get(0).getAddressLine(0);
            View marker = ((LayoutInflater) getSystemService(Context.LAYOUT_INFLATER_SERVICE)).inflate(R.layout.custom_marker, null);
            TextView  markerText = (TextView ) marker.findViewById(R.id.markerText);
            markerText.setText(address);
            map.clear();
            LatLng gps = new LatLng(location.getLatitude(), location.getLongitude());
            map.addMarker(new MarkerOptions()
                    .position(gps)
                    .title("Current Position")
                    .icon(BitmapDescriptorFactory.fromBitmap(createDrawableFromView(this, marker))));
            map.animateCamera(CameraUpdateFactory.newLatLngZoom(gps, 12));
        }

    }

    public static Bitmap createDrawableFromView(Context context, View view) {
        DisplayMetrics displayMetrics = new DisplayMetrics();
        ((Activity) context).getWindowManager().getDefaultDisplay().getMetrics(displayMetrics);
        view.setLayoutParams(new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT));
        view.measure(displayMetrics.widthPixels, displayMetrics.heightPixels);
        view.layout(0, 0, displayMetrics.widthPixels, displayMetrics.heightPixels);
        view.buildDrawingCache();
        Bitmap bitmap = Bitmap.createBitmap(view.getMeasuredWidth(), view.getMeasuredHeight(), Bitmap.Config.ARGB_8888);

        Canvas canvas = new Canvas(bitmap);
        view.draw(canvas);

        return bitmap;
    }

    private void getCurrentLocation() throws SecurityException, IOException {//TODO добавить рантайм проверку
        boolean isGPSEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
        boolean isNetworkEnabled = locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER);

        Location location = null;
        if (!(isGPSEnabled || isNetworkEnabled))
            Snackbar.make(mapView, R.string.error_location, Snackbar.LENGTH_INDEFINITE).show();
        else {
            if (isNetworkEnabled) {
                locationManager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER,
                        Config.LOCATION_UPDATE_MIN_TIME, Config.LOCATION_UPDATE_MIN_DISTANCE, mLocationListener);
                location = locationManager.getLastKnownLocation(LocationManager.NETWORK_PROVIDER);
            }

            if (isGPSEnabled) {
                locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER,
                        Config.LOCATION_UPDATE_MIN_TIME, Config.LOCATION_UPDATE_MIN_DISTANCE, mLocationListener);
                location = locationManager.getLastKnownLocation(LocationManager.GPS_PROVIDER);
            }
        }
        if (location != null) {
            Toast.makeText(getApplicationContext(), String.format("getCurrentLocation(%f, %f)", location.getLatitude(),
                    location.getLongitude()), Toast.LENGTH_SHORT).show();
            drawMarker(location);
        }
    }

    public void showSearchDialog() {
        ChooseDirectionDialog dialog = new ChooseDirectionDialog();
        dialog.show(getSupportFragmentManager(), "SEARCH_DIALOG_TAG");
    }

}

File: app/src/main/java/ru/taximer/taxiandroid/ui/MainActivity.kt
package ru.taximer.taxiandroid.ui


import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.location.Address
import android.location.Geocoder
import android.location.Location
import com.google.android.gms.location.LocationListener
import android.location.LocationManager
import android.support.design.widget.BottomSheetBehavior
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.RecyclerView
import android.util.DisplayMetrics
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import android.widget.TextView
import android.widget.Toast
import android.widget.RelativeLayout.LayoutParams
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationServices
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapView
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Marker
import com.google.android.gms.maps.model.MarkerOptions
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_main.main_container
import kotlinx.android.synthetic.main.activity_main.mapView
import kotlinx.android.synthetic.main.custom_marker.markerText
import kotlinx.android.synthetic.main.search_location_sheeet.autocompleteAddresses
import kotlinx.android.synthetic.main.search_location_sheeet.tmpBar

import java.io.IOException
import java.util.Locale

import ru.taximer.taxiandroid.R
import ru.taximer.taxiandroid.network.Config
import ru.taximer.taxiandroid.network.models.PlaceLocationModel
import ru.taximer.taxiandroid.utils.PERMISSIONS_LOCATION
import ru.taximer.taxiandroid.utils.arePermissionsGranted
import ru.taximer.taxiandroid.utils.isNotNullOrEmpty
import ru.taximer.taxiandroid.utils.shouldShowRationale

class MainActivity : AppCompatActivity(), OnMapReadyCallback, LocationListener {

    var map: GoogleMap? = null
    var mBottomSheetBehavior: BottomSheetBehavior<*>? = null
    var currentMarker: Marker? = null


    private var locationProvider: FusedLocationProviderClient? = null
    private var addressDisposable: Disposable? = null
    private var isDetectedPlace = true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        mapView.onCreate(savedInstanceState)
        mapView.getMapAsync(this)

        val sheetView = autocompleteAddresses.parent.parent as FrameLayout
        mBottomSheetBehavior = BottomSheetBehavior.from(sheetView)
        if (mBottomSheetBehavior != null) {
            mBottomSheetBehavior!!.setBottomSheetCallback(object : BottomSheetBehavior.BottomSheetCallback() {
                override fun onStateChanged(bottomSheet: View, newState: Int) {}

                override fun onSlide(bottomSheet: View, slideOffset: Float) {}
            })
        }

        mBottomSheetBehavior!!.peekHeight = 130
        tmpBar.requestLayout()
    }

    override fun onResume() {
        super.onResume()
        mapView.onResume()
    }

    override fun onPause() {
        super.onPause()
        mapView.onPause()
    }

    override fun onBackPressed() {
        if (mBottomSheetBehavior!!.state != BottomSheetBehavior.STATE_HIDDEN) {
            mBottomSheetBehavior!!.setState(BottomSheetBehavior.STATE_COLLAPSED)
        }
        else {
            super.onBackPressed()
        }
    }

    override fun onDestroy() {
        addressDisposable?.dispose()
        addressDisposable = null

        super.onDestroy()
    }

    ///////////////////////////////////////////////////////////////////////////
    // Map stuff
    ///////////////////////////////////////////////////////////////////////////

    override fun onMapReady(googleMap: GoogleMap) {
        map = googleMap
        map?.isIndoorEnabled = true
        checkMyLocationPermissionSilent {
            allowMyLocation()
        }
        getCurrentLocation()

        map?.moveCamera(CameraUpdateFactory.zoomTo(11f))
    }

    @SuppressLint("MissingPermission")
    private fun allowMyLocation() {
        map?.isMyLocationEnabled = true
        map?.uiSettings?.isMyLocationButtonEnabled = true
    }

    private fun getCurrentLocation() {
        checkMyLocationPermission {
            locationProvider = LocationServices.getFusedLocationProviderClient(this)
            locationProvider?.lastLocation?.addOnSuccessListener {
                onLocationChanged(it)
            }
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    // Location permission flow
    ///////////////////////////////////////////////////////////////////////////

    private fun checkMyLocationPermission(func: () -> Unit) {
        if (arePermissionsGranted(*PERMISSIONS_LOCATION)) func()
        else showLocationRationale()
    }

    private fun checkMyLocationPermissionSilent(func: () -> Unit) {
        if (arePermissionsGranted(*PERMISSIONS_LOCATION)) func()
    }

    private fun showLocationRationale() {
        if (shouldShowRationale(*PERMISSIONS_LOCATION)) {
            val snackbar = Snackbar.make(main_container, getText(R.string.rationale_location_permission), Snackbar.LENGTH_LONG)
            snackbar.setAction(R.string.label_ok) {
                requestLocationPermission()
            }
            snackbar.show()
        }
        else requestLocationPermission()
    }

    private fun requestLocationPermission() {
        requestPermissions(
                PERMISSIONS_LOCATION,
                PERMISSION_REQUEST_CODE_LOCATION
        )
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        when (requestCode) {
            PERMISSION_REQUEST_CODE_LOCATION -> {
                if (arePermissionsGranted(*PERMISSIONS_LOCATION)) {
                    allowMyLocation()
                    getCurrentLocation()
                }
                else {
                    val snackbar = Snackbar.make(main_container, getText(R.string.rationale_location_permission), Snackbar.LENGTH_LONG)
                    snackbar.setAction(R.string.label_ok) {
                        requestLocationPermission()
                    }
                    snackbar.show()
                }
            }
            else -> super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    // Location change listener
    ///////////////////////////////////////////////////////////////////////////

    override fun onLocationChanged(location: Location?) {
        location ?: return
        drawMarker(location)
    }

    @Throws(IOException::class)
    private fun drawMarker(location: Location?) {
        location ?: return
        val gps = LatLng(location.longitude, location.latitude)
        map?.let {
            detectAddress(gps)
            if(currentMarker != null){
                currentMarker?.position = gps
            }else{
                val marker = (getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater).inflate(R.layout.custom_marker, null)
                it.addMarker(MarkerOptions()
                        .position(gps)
                        .title("Current Position")
                        .icon(BitmapDescriptorFactory.fromBitmap(createDrawableFromView(this, marker))))
                it.animateCamera(CameraUpdateFactory.newLatLngZoom(gps, DEFAULT_ZOOM))
            }
        }

    }

    private fun detectAddress(coordinates: LatLng) {
        addressDisposable?.dispose()

        addressDisposable = Observable.fromCallable<PlaceLocationModel> {
            val geocoder = Geocoder(this, Locale.getDefault())
            val addresses = geocoder.getFromLocation(coordinates.latitude, coordinates.longitude, 1)
            getPlaceLocationModel(addresses, coordinates)
        }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { markerText.text = it.address},
                        { detectFallbackReverseGeocode(coordinates) })
    }

    private fun detectFallbackReverseGeocode(coordinates: LatLng) {
        addressDisposable?.dispose()

        addressDisposable = Observable.create(FallbackReverseGeocoderObservable(coordinates.latitude, coordinates.longitude))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread()).subscribe(
                        {
                            val place = getPlaceLocationModel(it, coordinates)
                            markerText.text = place.address

                        },
                        {
                            it.printStackTrace()
                            val place = PlaceLocationModel(coordinates.latitude, coordinates.longitude, "nope")
                            markerText.text = place.address
                        })
    }

    private fun getPlaceLocationModel(addresses: List<Address>, coordinates: LatLng): PlaceLocationModel {
        val fullAddress = StringBuffer()
        if (addresses.isNotNullOrEmpty()) {
            for (line in 0..addresses[0].maxAddressLineIndex) {
                fullAddress.append(addresses[0].getAddressLine(line) ?: "")
                if (line != addresses[0].maxAddressLineIndex) fullAddress.append(", ")
            }
        }
        return PlaceLocationModel(coordinates.latitude, coordinates.longitude, fullAddress.toString())
    }

    fun showSearchDialog() {
        val dialog = ChooseDirectionDialog()
        dialog.show(supportFragmentManager, "SEARCH_DIALOG_TAG")
    }

    companion object {

        private const val PERMISSION_REQUEST_CODE_LOCATION = 7
        private const val DEFAULT_ZOOM = 12f

        fun createDrawableFromView(context: Context, view: View): Bitmap {
            val displayMetrics = DisplayMetrics()
            (context as Activity).windowManager.defaultDisplay.getMetrics(displayMetrics)
            view.layoutParams = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
            view.measure(displayMetrics.widthPixels, displayMetrics.heightPixels)
            view.layout(0, 0, displayMetrics.widthPixels, displayMetrics.heightPixels)
            view.buildDrawingCache()
            val bitmap = Bitmap.createBitmap(view.measuredWidth, view.measuredHeight, Bitmap.Config.ARGB_8888)

            val canvas = Canvas(bitmap)
            view.draw(canvas)

            return bitmap
        }
    }

}
